### PR TITLE
[next] salt: backport upstream version.py implementation

### PR DESCRIPTION
**Pull request by amstewart**
In nanbield, salt compilation fails, with setuptools throwing an error that that salt version generated by salt/version.py is invalid. Backport the upstream-master implementation of the file, which complies with modern PEP-version expectations.

This change is fundamentally a response to the setuptools version in nanbield's `setuptools-legacy.bbclass` implementation upgrading.

[AB#2467374](https://dev.azure.com/ni/94b22d7b-ad7b-4f5e-88f0-867910f91c94/_workitems/edit/2467374)

# Testing
* [x] The `salt` recipe now build successfully on the /next branch.

### CODE REVIEW

File: recipes-support/salt/files/0001-version.py-merge-upstream-version.py-implementation.patch [hunk 1]
```diff
@@ -0,0 +1,1370 @@
...
+like...
+```
+|     raise InvalidVersion(f"Invalid version: '{version}'")
+| setuptools.extern.packaging.version.InvalidVersion: Invalid version: '3000.2-32-gfa9550c7f3'
+```
+
+Backport upstream saltstack/salt:master's verison.py implementation,
+which has been refactored to generate compliant version strings.
+
+Upstream-Status: Submitted [https://github.com/ni/salt/pull/69]
```
Comment (chaitu236) [inline recipes-support/salt/files/0001-version.py-merge-upstream-version.py-implementation.patch:19]: I assume we'd need to revert this change once https://github.com/ni/salt/pull/69 gets merged?

**Review by chaitu236 — COMMENTED** [review]
Why `nilrt/master/next` and not `nilrt/master/nanbield`?

**Comment by amstewart** [comment]
> Why `nilrt/master/next` and not `nilrt/master/nanbield`?

Because we're not actually shipping a nanbield release. Once we get to C2, OE will release 'scarthgap', which is what we actually want to release with. This mainline exists as the evergreen "next" major release.

File: recipes-support/salt/files/0001-version.py-merge-upstream-version.py-implementation.patch (hunk 1 above)
Comment (amstewart) [reply recipes-support/salt/files/0001-version.py-merge-upstream-version.py-implementation.patch:19]: It has been merged. So I'm going to drop this PR.

**Comment by amstewart** [comment]
CLOSED: The upstream ni/salt PR has been closed. So there is no need for us to apply our own .patch.
